### PR TITLE
chore: add a separate go.mod for tools versioning

### PR DESCRIPTION
**Pull request by pmalek**
**What this PR does / why we need it**:

In order to properly keep tabs on tools versions. Up until now we've been using the [`go-get-tool`](https://github.com/Kong/kubernetes-ingress-controller/blob/7d5108a9baf66d28ed737de78ff29b81db48aa44/Makefile#L46-L58) Makefile definition which has its flaws (e.g. it doesn't download a tool whenever a corresponding `bin/<tool>` file exists regardless of the version that it has been invoked with).

Instead this PR proposes to use go toolchain + a separate `go.mod` file in order to track tools versions.

With this proposal the procedure to download a particular tool would be changed to something like:

* add a new tool or change an existing tool's version in `third_party/go.mod`
* run below mentioned command (swapping golangci-lint for the tool that you're adding/updating)
 
  ```
  (cd third_party && go mod tidy ) && GOBIN=$(pwd)/bin go install -modfile third_party/go.mod github.com/golangci/golangci-lint/cmd/golangci-lint
  ```

This PR also bumps `k8s.io/code-generator` to `v0.24.0` because that's what go toolchain allows (possibly due to other modules dependencies requirements (if that's a blocker and we'd like to have separate dependencies trees for all tools then we'd have to rethink the organization of directories - possibly putting every tool in a separate module).

### Open questions

* ~~`tools/` dir placement (and name) is open for suggestions`~~

  We decided to move forward with `third_party`

### CODE REVIEW

**Review by shaneutt — COMMENTED** [review]
Just to confirm, ultimately what we want is this:

1. dependabot can automatically keep our build tools up to date
2. when devs pull dependabot updates from main, subsequent makefile target calls pull the newest version if applicable (whereas currently it wont update if the file exists)

Correct?

**Comment by pmalek** [comment]
> 1. dependabot can automatically keep our build tools up to date

Yes.

> 2. when devs pull dependabot updates from main, subsequent makefile target calls pull the newest version if applicable (whereas currently it wont update if the file exists)

Yes but not only for newer versions. It will work for any version that one would put in tools' `go.mod`, contrary to what happens now in such a scenario (which is just a plain check for file's existence).

File: Makefile
Comment (shaneutt) [inline Makefile:35]: :+1:

File: Makefile
Comment (shaneutt) [inline Makefile:134]: :+1:

**Review by shaneutt — CHANGES_REQUESTED** [review]
On the note about `tools/` vs other locations: if we're all assuming [golang-standards/project-layout](https://github.com/golang-standards/project-layout) as our de-facto influence I don't think `tools/` fits well as repositories like [Istio](https://github.com/istio/istio/tree/master/tools) generally use this for scripts and go programs that are ancillary to the main project, but **are developed in the same repository** which doesn't match what we're trying to do.

Looking for an alternative I narrowed it down to `build/` and `third_party/` seeming to make the most sense. At first I thought `build/` makes the most sense of these two because "building" is inherently the reason we want many of these tools (e.g. `kustomize` for building deploy manifests, `client-gen` for building client code, e.t.c.) but there's also things in here like `golangci-lint` which are a little less clear. The purpose of the "tools" we're going to manage isn't unified or distinct, it's kinda nebulous, and from what I can tell that matches with `third_party`. It's kinda a nebulous place to put tools maintained elsewhere:

```
External helper tools, forked code and other 3rd party utilities (e.g., Swagger UI).
```

So I'm gonna vote for `third_party/` for now, I think it's reasonable.

So mostly LGTM just before we merge if you agree with my assessment about the directory let's change that (the good news is, this is all easy to change later) and then there was a _surprising_ number of client generator changes... if you would just trigger an E2E test run on this PR before we merge I would appreciate it, just a little bit more coverage.

File: Makefile
Comment (jrsmroz) [reply Makefile:35]: Dapr imports external tools https://github.com/dapr/dapr/blob/master/tools/tools.go

Comment (shaneutt) [reply Makefile:35]: If you feel strongly about using `tools/` I could be persuaded, but if you're on board for `third_party` I'm just more used to seeing repositories that use `tools/` for tools where the source is maintained therein.

**Comment by pmalek** [comment]
> On the note about `tools/` vs other locations: if we're all assuming [golang-standards/project-layout](https://github.com/golang-standards/project-layout) as our de-facto influence I don't think `tools/` fits well as repositories like [Istio](https://github.com/istio/istio/tree/master/tools) generally use this for scripts and go programs that are ancillary to the main project, but **are developed in the same repository** which doesn't match what we're trying to do.
> 
> Looking for an alternative I narrowed it down to `build/` and `third_party/` seeming to make the most sense. At first I thought `build/` makes the most sense of these two because "building" is inherently the reason we want many of these tools (e.g. `kustomize` for building deploy manifests, `client-gen` for building client code, e.t.c.) but there's also things in here like `golangci-lint` which are a little less clear. The purpose of the "tools" we're going to manage isn't unified or distinct, it's kinda nebulous, and from what I can tell that matches with `third_party`. It's kinda a nebulous place to put tools maintained elsewhere:
> 
> ```
> External helper tools, forked code and other 3rd party utilities (e.g., Swagger UI).
> ```
> 
> So I'm gonna vote for `third_party/` for now, I think it's reasonable.
> 
> So mostly LGTM just before we merge if you agree with my assessment about the directory let's change that (the good news is, this is all easy to change later) and then there was a _surprising_ number of client generator changes... if you would just trigger an E2E test run on this PR before we merge I would appreciate it, just a little bit more coverage.

So I didn't really have a strong opinion on this but given your explanation I agree with using `third_party`. Especially since we might end up in the future needing to use `tools/` for our own tools. It would be relatively easy to move what's being added in this PR but still. I think it fits better 👍 

One thing though about the e2e tests is that it seems they started to consistently fail recently hence I propose to wait until that is resolved. I've created a separate issue for that: https://github.com/Kong/kubernetes-ingress-controller/issues/2610.

Until that resolution I'm adding a `do-not-merge` label.

**Comment by shaneutt** [comment]
https://github.com/Kong/kubernetes-ingress-controller/issues/2610 is set we should be good to merge when ready now.

I also created https://github.com/Kong/gateway-operator/issues/47 to have this done for the GWO as well.

**Review by shaneutt — DISMISSED** [review]

**Comment by pmalek** [comment]
I've run e2e tests on this one  and they pass all but one: https://github.com/Kong/kubernetes-ingress-controller/runs/7088474940?check_suite_focus=true#step:5:1215

```
all_in_one_test.go:364: 
        	Error Trace:	all_in_one_test.go:364
        	Error:      	Condition never satisfied
        	Test:       	TestDeployAllInOnePostgresWithMultipleReplicas
=== CONT  TestDeployAllInOneEnterprisePostgres
    all_in_one_test.go:387: building test cluster and environment
--- FAIL: TestDeployAllInOnePostgresWithMultipleReplicas (263.01s)
time="2022-06-28T08:30:45Z" level=error msg="checking config status failed" error="making HTTP request: Get \"[http://172.18.0.241:8001/status\](http://172.18.0.241:8001/status/)": dial tcp 172.18.0.241:8001: connect: no route to host"
```

🤔 

https://github.com/Kong/kubernetes-ingress-controller/actions/runs/2574755030

**Comment by shaneutt** [comment]
> I've run e2e tests on this one and they pass all but one - [e2e-tests (v1.22.7, v1.11.8)](https://github.com/Kong/kubernetes-ingress-controller/runs/7088474940?check_suite_focus=true#logs)
> 
> ```
> all_in_one_test.go:364: 
>         	Error Trace:	all_in_one_test.go:364
>         	Error:      	Condition never satisfied
>         	Test:       	TestDeployAllInOnePostgresWithMultipleReplicas
> === CONT  TestDeployAllInOneEnterprisePostgres
>     all_in_one_test.go:387: building test cluster and environment
> --- FAIL: TestDeployAllInOnePostgresWithMultipleReplicas (263.01s)
> time="2022-06-28T08:30:45Z" level=error msg="checking config status failed" error="making HTTP request: Get \"[http://172.18.0.241:8001/status\](http://172.18.0.241:8001/status/)": dial tcp 172.18.0.241:8001: connect: no route to host"
> ```
> 
> thinking
> 
> https://github.com/Kong/kubernetes-ingress-controller/actions/runs/2574755030



> I've run e2e tests on this one and they pass all but one - [e2e-tests (v1.22.7, v1.11.8)](https://github.com/Kong/kubernetes-ingress-controller/runs/7088474940?check_suite_focus=true#logs)
> 
> ```
> all_in_one_test.go:364: 
>         	Error Trace:	all_in_one_test.go:364
>         	Error:      	Condition never satisfied
>         	Test:       	TestDeployAllInOnePostgresWithMultipleReplicas
> === CONT  TestDeployAllInOneEnterprisePostgres
>     all_in_one_test.go:387: building test cluster and environment
> --- FAIL: TestDeployAllInOnePostgresWithMultipleReplicas (263.01s)
> time="2022-06-28T08:30:45Z" level=error msg="checking config status failed" error="making HTTP request: Get \"[http://172.18.0.241:8001/status\](http://172.18.0.241:8001/status/)": dial tcp 172.18.0.241:8001: connect: no route to host"
> ```
> 
> thinking
> 
> https://github.com/Kong/kubernetes-ingress-controller/actions/runs/2574755030

Recommend running locally and if it's reproducible freeze the tests at the failure point and inspect the cluster state.

**Comment by shaneutt** [comment]
What was the source of that other failure?

**Comment by pmalek** [comment]
> What was the source of that other failure?

```
"2.8.1" does not contain "enterprise-edition"
```

Related PR: https://github.com/Kong/kubernetes-ingress-controller/pull/2615

**Comment by pmalek** [comment]
> What was the source of that other failure?

Alright, so I've [added some more logs to the failing e2e test](https://github.com/Kong/kubernetes-ingress-controller/pull/2619) - `TestDeployAllInOnePostgresWithMultipleReplicas` - and it seems that for some reason there's a metric missing - specifically `ingress_controller_configuration_push_count`. I'm not sure why and I'm not sure if there's a pattern in the recent e2e test failures (in the test runs that I've triggered).

example of a failure: https://github.com/Kong/kubernetes-ingress-controller/runs/7100064759?check_suite_focus=true#step:5:1389

**Comment by pmalek** [comment]
It seems that this is already failing on `main` so most likely not related to this PR: [today's CI failure on main](https://github.com/Kong/kubernetes-ingress-controller/runs/7105575038?check_suite_focus=true#step:5:195)

**Comment by shaneutt** [comment]
> Would we prefer to sort out the [e2e failure](https://github.com/Kong/kubernetes-ingress-controller/issues/2622) first to reliable test this PR or do we feel comfortable with the statement that it's unrelated to the changes introduced here?

I'm having a really hard time seeing how this would related, I think we can move forward assuming that it's unrelated. On top of that keep in mind that the E2E tests is to catch weird edges so it's not the end of the world (and in fact it's how we've set it up) if it ends up being related, the intention is that we wont release until it's resolved.

**Comment by pmalek** [comment]
> I'm having a really hard time seeing how this would relate, I think we can move forward assuming that it's unrelated. On top of that keep in mind that the E2E tests is to catch weird edges so it's not the end of the world (and in fact it's how we've set it up) if it ends up being related, the intention is that we wont release until it's resolved.

I had the same view on this but just wanted to be sure that it makes sense to dismiss the failure from the PoV of this PR.  PTAL then, re-requested a review

**Review by shaneutt — APPROVED** [review]

**Comment by shaneutt** [comment]
Thanks for adding this, this was very forward-thinking and I'm excited to see it in the gateway operator as well :vulcan_salute:

**Pull request merged**
